### DISCUSSION TITLE
Import ESM module

### DISCUSSION THREAD

**Discussion by wimurk** [adonisjs/core#3620] [Help] [answered]
## Package version
@latest

## Node.js and npm version
@latest

## Sample Code (to reproduce the issue)
`import clipboard from 'clipboardy';`

## BONUS (a sample repo to reproduce the issue)
Install the package `clipboardy`. Then import it

I try to import the above package. But it gives me the error
```typescript
 require() of ES Module ......\node_modules\clipboardy\index.js from .......app\Controllers\Http\RobotsController.ts not supported.
Instead change the require of index.js in .....app\Controllers\Http\RobotsController.ts to a dynamic import() which is available in all CommonJS modules.

 at Object.<anonymous> ......app/Controllers/Http/RobotsController.ts:9
   4|   import axios from 'axios';
   5|   import https from 'https';
   6|   import robot from 'robotjs';
   7|   import {keyboard, Key} from '@nut-tree/nut-js';
   8|   import execa from 'execa';
>  9|   import clipboard from 'clipboardy';

```

I have tried dynamic import, async import, sync import. nothing works and i  cant find anything on import ESM into adonis. It works with other node projects. Just not with adonis

**Comment by Julien-R44** [comment, marked as answer]
See here for more informations : https://news.adonisjs.com/issues/round-up-2021-and-the-future-848097

> **Moving to ESM**
Slowly the Node.js ecosystem is adopting ES modules over CommonJS. However, the road to ES modules is a little bumpy, and also, TypeScript has some [pending issues](https://github.com/microsoft/TypeScript/issues/46334?utm_campaign=Monthly%20Updates&utm_medium=email&utm_source=Revue%20newsletter) before AdonisJS can consider moving to ESM.
I see myself documenting the migration process somewhere around April 2022 and have an open window for at-least 6months for the projects to adopt ESM.

In the meantime, you can just use the v2.3.0 version of clipboardy.

**Reply by wimurk** [reply]
I have been searching for hours and didn't read that particular part. I will stick with 2.3.0 until the update release. THanks for the fast reply

**Reply by IARI** [reply]
the link gives me a 522 server error - is there any updated info on the subject?

**Reply by McSneaky** [reply]
> the link gives me a 522 server error - is there any updated info on the subject?

Seems to be down yeah. Thing is, that whole NodeJS ecosystem is slowly moving from CommonJS require over to ESM imports. Adonis will support ESM out-of box with version 6, which is currently in works

Meanwhile you have to keep using package versions that support CommonJS

**Reply by wavedeck** [reply]
IMHO the time when Adonis 6 comes out with native ESM support, will most likely be the rise of AdonisJS for production applications. 

Most companies use NestJS and they have the exact same issues of not supported ESM, while libraries are rapidly switching to ESM only.

I‘m really hoping for v6 anytime this year, so i can switch my applications from being custom fastify apps to Adonis v6.

 If necessary I‘m totally open for PR‘s where work is still needed to finally see ESM in big production ready frameworks happen 🥳

**Reply by thetutlage** [reply]
Its happening this year for sure. Infact before winters

**Comment by angeld287** [comment]
In my case I had the same problem in node js with node-fetch. I solved it using the dynamic import:

`import { RequestInfo, RequestInit } from 'node-fetch';`

`const fetch = (url: RequestInfo, init?: RequestInit) =>`
    `import('node-fetch').then(({ default: fetch }) => fetch(url, init));`

`export default fetch`
